platforms/core-execution/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -200,6 +217,16 @@ public void record(FileSystemLocationSnapshot snapshot) {
         virtualFileSystem.store(snapshot.getAbsolutePath(), () -> snapshot);
     }
 
+    @Override
+    public void moveAtomically(String sourceLocation, String targetLocation) throws IOException {
+        FileSystemLocationSnapshot sourceSnapshot = read(sourceLocation);
+        write(ImmutableList.of(sourceLocation, targetLocation), () -> {
+            Files.move(Paths.get(sourceLocation), Paths.get(targetLocation), ATOMIC_MOVE);
```

Comment:
❓ How is ATOMIC_MOVE supported accross filesystems?

Reply:
That's a good question, since the whole operation hinges on this. :) FTR, we don't necessarily need _atomic_ moves; we just need a rename to happen instead of a file-by-file copy. And this seems to be what `ATOMIC_MOVE` ensures.

On Windows this is handled via `MoveFileEx`:

https://github.com/frohoff/jdk8u-jdk/blob/da0da73ab82ed714dc5be94acd2f0d00fbdfe2e9/src/windows/classes/sun/nio/fs/WindowsFileCopy.java#L290-L304

Crucially, this is executed without `MOVEFILE_COPY_ALLOWED`, which would fail if the source and target are on different file systems, but this is not the case with what we are asking for:

https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-movefileexa#parameters

On Unixes it's done via `rename(2)` which also fails if you try to rename paths between file systems, but otherwise should work fine, even if the target exists and needs to be replaced:

> If newpath already exists, it will be atomically replaced, so
> that there is no point at which another process attempting to
> access newpath will find it missing.

https://man7.org/linux/man-pages/man2/rename.2.html

The man page for `rename(2)` on macOS Sonoma says:

> The rename() system call guarantees that an instance of new will always exist, even if the system should crash in the middle of the operation.

Reply:
All-in-all I expect that we'll be fine.

Reply:
Ok, we will anyway soon see if something once we have the change out there for user to check.